src/main/java/org/prebid/server/bidder/improvedigital/ImprovedigitalBidder.java

```diff
@@ -71,16 +100,37 @@ public ImprovedigitalBidder(String endpointUrl, JacksonMapper mapper) {
         return bidsFromResponse(bidRequest, bidResponse);
     }
 
-    private static List<BidderBid> bidsFromResponse(BidRequest bidRequest, BidResponse bidResponse) {
+    private List<BidderBid> bidsFromResponse(BidRequest bidRequest, BidResponse bidResponse) {
         return bidResponse.getSeatbid().stream()
                 .filter(Objects::nonNull)
                 .map(SeatBid::getBid)
                 .filter(Objects::nonNull)
                 .flatMap(Collection::stream)
-                .map(bid -> BidderBid.of(bid, getBidType(bid.getImpid(), bidRequest.getImp()), bidResponse.getCur()))
+                .map(bid -> makeBidderBid(bidRequest, bidResponse, bid))
+                .filter(Objects::nonNull)
                 .collect(Collectors.toList());
     }
 
+    private BidderBid makeBidderBid(BidRequest bidRequest, BidResponse bidResponse, Bid bid) {
+        Bid modifiedBid = bid;
+        if (bid.getExt() != null) {
+            final ImprovedigitalBidExt improvedigitalBidExt;
+            try {
+                improvedigitalBidExt = mapper.mapper().treeToValue(bid.getExt(), ImprovedigitalBidExt.class);
+            } catch (JsonProcessingException e) {
+                throw new PreBidException(e.getMessage(), e);
+            }
+            // Populate dealId
+            String buyingType = improvedigitalBidExt.getImprovedigital().getBuyingType();
+            Integer lineItemId = improvedigitalBidExt.getImprovedigital().getLineItemId();
+            if (!StringUtils.isBlank(buyingType) && buyingType.matches("(classic|deal)")
+                    && lineItemId != null && lineItemId > 0) {
+                modifiedBid = bid.toBuilder().dealid(lineItemId.toString()).build();
+            }
+        }
+        return BidderBid.of(modifiedBid, getBidType(bid.getImpid(), bidRequest.getImp()), bidResponse.getCur());
+    }
```

Comment:
This can be refactored to this
```java
private List<BidderBid> bidsFromResponse(BidRequest bidRequest, BidResponse bidResponse) {
        return bidResponse.getSeatbid().stream()
                .filter(Objects::nonNull)
                .map(SeatBid::getBid)
                .filter(Objects::nonNull)
                .flatMap(Collection::stream)
                .map(bid -> BidderBid.of(resolveBid(bid), getBidType(bid.getImpid(),
                        bidRequest.getImp()), bidResponse.getCur()))
                .collect(Collectors.toList());
    }

    private Bid resolveBid(Bid bid) {
        final JsonNode bidExt = bid.getExt();
        if (bidExt != null && !bidExt.isEmpty()) {
            final ImprovedigitalBidExtImprovedigital improvedigitalBidExt;
            try {
                improvedigitalBidExt = mapper.mapper().treeToValue(bid.getExt().get("improvedigital"),
                        ImprovedigitalBidExtImprovedigital.class);
            } catch (JsonProcessingException e) {
                throw new PreBidException(e.getMessage());
            }
            // Populate dealId
            final String buyingType = improvedigitalBidExt.getBuyingType();
            final Integer lineItemId = improvedigitalBidExt.getLineItemId();
            if (StringUtils.isNotBlank(buyingType)
                    && buyingType.matches("(classic|deal)")
                    && lineItemId != null
                    && lineItemId > 0) {

                return bid.toBuilder().dealid(lineItemId.toString()).build();
            }
        }

        return bid;
    }
```
After that ImprovedigitalBidExt class is no longer needed and ImprovedigitalBidExtImprovedigital class can take better name

Reply:
I believe whether ImprovedigitalBidExt is defined or `improvedigital` is hard-coded in `bid.getExt().get("improvedigital")` is a matter of personal taste (as can be seen in different approaches picked in different adapters). I chose to follow the code from Appnexus bidder adapter in defining however trivial ImprovedigitalBidExt to avoid hard-coded literals like "improvedigital".

Reply:
@snahornyi thanks for the comments :-). I'll make changes and add more tests. Please see replies to some of your comments inline.

Reply:
Ok, agree with you)

Reply:
I made some change